File: Java/636.java
__________________________________________________________________________________________________
sample 5 ms submission
class Solution {
    public int[] exclusiveTime(int n, List<String> logs) {
        int[] res = new int[n];
        
        int start = 0;
        Stack<Integer> stack = new Stack<>();
        for (int i = 0, len = logs.size(); i < len; i++) {
            Log log = new Log(logs.get(i));

            if (log.isStart) {
                if (!stack.isEmpty()) {
                    res[stack.peek()] += log.time - start;
                }
                stack.push(log.id);
                start = log.time;
            } else {
                res[stack.pop()] += log.time - start + 1;
                start = log.time + 1;
            }
        }
        
        return res;
    }
}

class Log {
    int id;
    int time;
    boolean isStart;
    
    Log(String logStr) {
        int i = 0, len = logStr.length();
        id = 0;
        while (logStr.charAt(i) != ':') {
            id = id * 10 + (logStr.charAt(i++) - '0');
        }
        
        if (logStr.charAt(++i) == 's') {
            isStart = true;
            i += 6;
        } else {
            isStart = false;
            i += 4;
        }
        
        time = 0;
        while (i < len) {
            time = time * 10 + (logStr.charAt(i++) - '0');
        }
    }
}
__________________________________________________________________________________________________
sample 37004 kb submission
class Solution {
    public int[] exclusiveTime(int n, List<String> l) {
        List<Log> logs = l.stream().map(Log::new).collect(Collectors.toList());
        Stack<Integer> frames = new Stack<>();
        int curId = -1;
        int curStart = -1;
        int[] results = new int[n];
        for (Log log : logs) {
            if (log.isStart()) {
                if (curId != -1) {
                    results[curId] += log.time() - curStart;
                    frames.push(curId);
                }
                curId = log.id();
                curStart = log.time();
            }
            else {
                results[curId] += log.time() - curStart + 1;
                if (!frames.isEmpty()) {
                    curId = frames.pop();
                    curStart = log.time() + 1;
                }
                else {
                    curId = -1;
                    curStart = -1;
                }
            }
        }
        return results;
    }
    
    private class Log {
        private int id;
        private boolean isStart;
        private int time;
        
        public Log(String log) {
            String[] tokens = log.split(":");
            id = Integer.valueOf(tokens[0]);
            isStart = tokens[1].equals("start");
            time = Integer.valueOf(tokens[2]);
        }
        
        public int id() {
            return id;
        }
        
        public int time() {
            return time;
        }
        
        boolean isStart() {
            return isStart;
        }
        
        boolean isEnd() {
            return !isStart();
        }
    }
}
__________________________________________________________________________________________________

File: Python3/636.py
__________________________________________________________________________________________________
sample 72 ms submission
class Solution:
    def exclusiveTime(self, n: 'int', logs: 'List[str]') -> 'List[int]':
        res = [0] * n
        stack = []
   
        
        for log in logs:
            idx, start, time = log.split(':')
            idx, start, time = int(idx), start == 'start', int(time)
            if start:
                if stack:
                    res[stack[-1]] += time - last_time                    
                stack.append(idx)
                last_time = time
            else:
                stack.pop()
                res[idx] += time - last_time + 1
                last_time = time + 1
        return res                 
__________________________________________________________________________________________________
sample 12952 kb submission
from collections import deque

class Solution:
    def exclusiveTime(self, n: int, logs: List[str]) -> List[int]:
        stack = deque()         # task and total time accumulated so far and whether it is cut off
        times = [0] * n
        last_time = 0
        for log in logs:
            task, state, time = log.split(':')
            task = int(task)
            time = int(time)
            is_end = state == 'end'
            if not stack:
                stack.append([task, 0, False])
            elif is_end:
                times[task] += stack[-1][1] + time - last_time
                if not stack[-1][2]:
                    times[task] += 1
                del stack[-1]
            else: # start
                stack[-1][1] += time - last_time
                if stack[-1][2]:  # if cut off, last time is closing, which takes up 1, so subtract it
                    stack[-1][1] -= 1
                stack[-1][2] = True
                stack.append([task, 0, False])
            last_time = time
        return times
__________________________________________________________________________________________________
